### ISSUE TITLE
[consistent-type-imports] ignore imports marked as used

### ISSUE THREAD

**Issue by FloEdelmann** [typescript-eslint/typescript-eslint#4508]
### Description

Vue.js 3 has a syntax (`<script setup>`) that allows importing stuff in the JavaScript/TypeScript section, and using it in the HTML/template section:

```vue
<script setup lang="ts">
import { ref } from "vue";
import MyComponent from "./MyComponent.vue";

type ComponentType = typeof MyComponent extends new () => infer T ? T : never;

const myRef = ref<ComponentType>();
</script>

<template>
  <MyComponent ref="myRef" />
</template>
```

Here, `@typescript-eslint/consistent-type-imports` would report the `MyComponent` import, as it thinks it's only being used as a type. However, it is actually also used as a value in the template. So applying the autofix breaks the Vue build.

Vue + typescript-eslint users are likely to also have the [eslint-plugin-vue](https://github.com/vuejs/eslint-plugin-vue) installed. There, we have rules that mark variables and imports from the `<script setup>` section with ESLint's [`markVariableAsUsed` function](https://eslint.org/docs/developer-guide/working-with-rules#the-context-object) when they are used somewhere in the `<template>` section.

Would it be possible to ignore imports that have been marked with this function in `@typescript-eslint/consistent-type-imports`?

For reference, this problem has been reported in the `eslint-plugin-vue` repo first: https://github.com/vuejs/eslint-plugin-vue/issues/1784

### Fail

```vue
<script setup lang="ts">
import { ref } from "vue";
import MyComponent from "./MyComponent.vue"; // should indeed be a type import

type ComponentType = typeof MyComponent extends new () => infer T ? T : never;

const myRef = ref<ComponentType>();
</script>

<template>
  <div ref="myRef" />
</template>
```

### Pass

```vue
<script setup lang="ts">
import { ref } from "vue";
import MyComponent from "./MyComponent.vue"; // should not be a type import, as it's used in the <template>

type ComponentType = typeof MyComponent extends new () => infer T ? T : never;

const myRef = ref<ComponentType>();
</script>

<template>
  <MyComponent ref="myRef" />
</template>
```

**Comment by bradzacher** [comment]
I don't think that operating based on `eslintUsed` is a good idea because it could mean anything.
Literally any plugin that you have installed could set that flag to true. Meaning that all of a sudden there is a false negative which would be really hard to debug.

Ideally the vue parser should augment the scope with the additional information about the usages, rather than just marking the variable as used.

As an example - our scope analyser adds usages for `React` when JSX is found by augmenting the analysis, even though there isn't an actual `React` usage in the code.

**Comment by FloEdelmann** [comment]
@ota-meshi Maybe [`vue-eslint-parser`](https://github.com/vuejs/vue-eslint-parser) could do something like that? I don't know much about that package.

**Comment by ota-meshi** [comment]
I'll explain why later, but the conclusion is that it's very difficult to use ScopeManager in vue-eslint-parser to mark template variables.

**Comment by bradzacher** [comment]
> I'll explain why later, but the conclusion is that it's very difficult to use ScopeManager in vue-eslint-parser to mark template variables.

You already get the variables to mark them as used - couldn't you just synthesise and add references to the imports instead?

**Comment by ota-meshi** [comment]
vue-eslint-parser parses `<template>` and `<script>` as separate ASTs.
ESLint's core rules and other plugin rules only check `<script>`. This means that popular plugins that don't know the `<template>`'s own AST will work. For example, the `<template>` AST does not have an `esTreeNodeToTSNodeMap`, so trying to check the `<template>` AST with typescript-eslint's rules can cause a crash.
Since eslint-plugin-vue rules implements to check `<template>` in addition to that, `<template>` checking also works.

The ScopeManager only has the scope of `<script>` because the ASTs of `<template>` and `<script>` are separated. This is why it is difficult to mark variables in `<template>` in ScopeManager.
If we try to mark a variable in `<template>`, the ScopeManager will contain an Identifier that shouldn't be accessible, which may confuse some rules.

**Comment by ota-meshi** [comment]
> I don't think that operating based on eslintUsed is a good idea because it could mean anything.

I don't think it's a good idea as you say. However, I think it might be a good solution for users as a compromise.
The ESLint core prefer-const rule also seems to use eslintUsed.
https://github.com/eslint/eslint/blob/417191dff0dbfa353675c409e25f27f578ee1559/lib/rules/prefer-const.js#L174

**Comment by bradzacher** [comment]
> However, I think it might be a good solution for Vue users as a compromise.

I don't think we should subject all TS users to a hack just for Vue.
Probably worth making an extension rule within `eslint-plugin-vue`, if that's the workaround you want to use.

---

> the ScopeManager will contain an Identifier that shouldn't be accessible, which may confuse some rules.

FWIW - as mentioned, we already do some "weird" stuff in our scope manager for references to `React` from within JSX by adding a reference from itself to itself.

```ts
import React from 'react';
//     ^^^^^ Variable.references = [ Reference<Identifier, 'React', {line:1, column: 8}> ]

const x = <div />;
```
[playground with scope tree](https://typescript-eslint.io/play/#ts=4.5.2&jsx=true&sourceType=module&showAST=scope&code=JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4BYAKCrQgDsBneADzgF44AeAE2ADc4AegB8pIA&rules=N4XyA&tsConfig=N4XyA)

Relevant code that does this:
https://github.com/typescript-eslint/typescript-eslint/blob/5ab1d579b7a393601352325d81e3c8026953f58e/packages/scope-manager/src/referencer/Referencer.ts#L122-L141

https://github.com/typescript-eslint/typescript-eslint/blob/5ab1d579b7a393601352325d81e3c8026953f58e/packages/scope-manager/src/referencer/Referencer.ts#L103-L120

This code has been in place for 17 months without issue.
I don't think there's any reason vue couldn't do the same thing - add a read-only reference from the import to itself instead of marking it as `eslintUsed`.

**Comment by ota-meshi** [comment]
> FWIW - as mentioned, we already do some "weird" stuff in our scope manager for references to React from within JSX by adding a reference from itself to itself.

Oh. I didn't understand your first comment. Thank you for explaining in detail. I didn't have the idea to add a reference in the same place. I think it's worth trying.

> I don't think we should subject all TS users to a hack just for Vue.
Probably worth making an extension rule within eslint-plugin-vue, if that's the workaround you want to use.

Thank you for your opinion.  If the method of using ScopeManager doesn't work, I will consider creating an extension rule in eslint-plugin-vue.

**Comment by bradzacher** [comment]
I'll close this here, as you have a tracking issue in your repo.
Feel free to @ me if you need assistance or advice!
